lib/services/scraper/scraper.go

```diff
@@ -0,0 +1,148 @@
+package scraper
+
+import (
+	"fmt"
+	"net/url"
+
+	"go-google-scraper-challenge/helpers/log"
+	"go-google-scraper-challenge/lib/models"
+
+	"github.com/gocolly/colly"
+)
+
+type Scraper struct {
+	Result *models.Result
+}
+
+var selectors = map[string]string{
+	"wholePage":    "html",
+	"topImageAds":  "#tvcap .plantl a.pla-unit-title-link",
+	"topLinkAds":   "#tvcap .d5oMvf > a",
+	"sideImageAds": ".cu-container a.plantl.clickable-card",
+	"nonAds":       "#search .yuRUbf > a",
+	"bottomAds":    "#bottomads .uEierd > a",
+}
+
+const GoogleSearchUrl = "http://www.google.com/search?q=%s"
+
+func (s *Scraper) Run() error {
+	escapedKeyword := url.QueryEscape(s.Result.Keyword)
+	url := fmt.Sprintf(GoogleSearchUrl, escapedKeyword)
+
+	return s.startScraping(url)
+}
+
+func (s *Scraper) startScraping(url string) error {
+	collector := colly.NewCollector()
+
+	collector.OnRequest(s.requestHandler)
+	collector.OnResponse(s.responseHandler)
+	collector.OnError(s.errorHandler)
+
+	collector.OnHTML(selectors["wholePage"], s.wholePageCollector)
+
+	collector.OnHTML(selectors["nonAds"], func(e *colly.HTMLElement) {
+		s.addNonAdLinkToResult(e)
+	})
+
+	collector.OnHTML(selectors["topImageAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeImage, models.AdLinkPositionTop, e)
+	})
+
+	collector.OnHTML(selectors["topLinkAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeLink, models.AdLinkPositionTop, e)
+	})
+
+	collector.OnHTML(selectors["sideImageAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeImage, models.AdLinkPositionSide, e)
+	})
+
+	collector.OnHTML(selectors["bottomAds"], func(e *colly.HTMLElement) {
+		s.addAdLinkToResult(models.AdLinkTypeLink, models.AdLinkPositionBottom, e)
+	})
+
+	collector.OnScraped(s.finishScrapingHandler)
+
+	err := collector.Visit(url)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Scraper) requestHandler(request *colly.Request) {
+	userAgent := RandomUserAgent()
+	request.Headers.Set("User-Agent", userAgent)
+
+	log.Info("Visiting ", request.URL)
+
+	err := models.UpdateResultStatus(s.Result, models.ResultStatusProcessing)
+	if err != nil {
+		log.Error("Failed to process result:", err.Error())
+	}
+}
+
+func (s *Scraper) responseHandler(response *colly.Response) {
+	log.Info("Visited ", response.Request.URL)
+}
+
+func (s *Scraper) errorHandler(response *colly.Response, errResponse error) {
+	result := s.Result
+	err := models.UpdateResultStatus(s.Result, models.ResultStatusFailed)
```

Comment:
Should we extract the action about `updating result status` out of the service?
It looks better to let the caller out there dealing with that action.
Because this service will have only 1 action about scraping and it's easy to maintain.

I see that you're trying to maintain the start of the result by stamping the `processing` status to the result.
That's why you put that logic here.
But what if we update the state of that result as `processing` before calling this service? 😇

Reply:
✅ moved in [e410587](https://github.com/carryall/go-google-scraper-challenge/pull/104/commits/e4105876eb75866b564be1bb009bd8ec847d31c8)